napari/layers/utils/link_layers.py

```diff
@@ -0,0 +1,145 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Iterable
+
+if TYPE_CHECKING:
+    from napari.layers import Layer
+
+#: Record of already linked layers... to avoid duplicating callbacks
+#  in the form of (id(layer1), id(layer2), attribute_name)
+_LINKED: set[tuple[int, int, str]] = set()
+
+
+# attributes that make no sense to link, or error on changing
+# we may want to include colormap here, but in most cases, it works fine.
+PROHIBITED = {'thumbnail', 'status', 'name'}
+
+
+def _get_common_evented_attributes(
+    layers: Iterable['Layer'],
+    prohibited: set[str] = PROHIBITED,
+    with_private=False,
+) -> set[str]:
+    """Get the set of common, non-private evented attributes in ``layers``.
+
+    Not all layer events are attributes, and not all attributes have
+    corresponding events.  Here we get the set of valid, non-private attributes
+    that are both events and attributes for the provided layer set.
+
+    Parameters
+    ----------
+    layers : iterable
+        A set of layers to evaluate for attribute linking.
+    prohibited : set, optional
+        A set of attributes to exclude, by default PROHIBITED
+    private : bool, optional
+        include private attributes
+
+    Returns
+    -------
+    names : set of str
+        A set of attribute names that may be linked between ``layers``.
+    """
+    from inspect import ismethod
+
+    try:
+        first_layer = next(iter(layers))
+    except StopIteration:
+        raise ValueError("``layers`` iterable must have at least one layer")
+    common_events = set.intersection(*(set(lay.events) for lay in layers))
+    common_attrs = set.intersection(*(set(dir(lay)) for lay in layers))
+    if not with_private:
+        common_attrs = {x for x in common_attrs if not x.startswith("_")}
+    common = common_events & common_attrs - prohibited
+
+    # lastly, discard any method-only events (we just want attrs)
+    for attr in set(common_attrs):
+        # properties do not count as methods and will not be excluded
+        if ismethod(getattr(first_layer.__class__, attr, None)):
+            common.discard(attr)
+
+    return common
+
+
+def experimental_link_layers(
+    layers: Iterable['Layer'], attributes: Iterable[str] = ()
+):
+    """Link ``attributes`` between all layers in ``layers``.
+
+    This essentially performs the following operation:
+
+    .. code-block:: python
+
+       for lay1, lay2 in permutations(layers, 2):
+           for attr in attributes:
+              lay1.events.<attr>.connect(_set_lay2_<attr>)
+
+    Recursion is prevented by checking for value equality prior to setting.
+
+    Parameters
+    ----------
+    layers : Iterable[napari.layers.Layer]
+        The set of layers to link
+    attributes : Iterable[str], optional
```

Comment:
I wonder maybe here if we can use `include` and `exclude` instead of `attributes` - thinking again about the pydantic [serialization/ to dict API](https://pydantic-docs.helpmanual.io/usage/exporting_models/#modeldict) which I quite like. If both `None` then get all common by default.

Reply:
actually... for the function that the user calls, I'm not 100% sure.
calling `link_layers([layer1, layer2], attributes=('contrast_limits','gamma'))` reads (to me) clearer than `include=('contrast...')`.  The user may not know anything about the pydantic API and serialization, so I'm not sure how much that parity matters here?

Reply:
do we want an `exclude` here? I think if we want both include and exclude then those words are nice. If we never want exclude then maybe attributes is fine. Let's get a third opinion from @jni. I would describe my preference here as weak, so happy with whatever

Reply:
I agree with that if we include an `exclude` argument.

Reply:
let's leave with attributes, i don't think `exclude` is critical

Reply:
I prefer attributes — I don't see much use for exclude here.